cpp/src/text/generate_ngrams.cu

```diff
@@ -0,0 +1,162 @@
+        return cudf::make_empty_column(cudf::data_type{cudf::STRING});
+
+    auto execpol = rmm::exec_policy(stream);
+    auto strings_column = cudf::column_device_view::create(strings.parent(),stream);
+    auto d_strings = *strings_column;
+
+    // first create a new offsets vector removing nulls and empty strings from the input column
+    std::unique_ptr<cudf::column> non_empty_offsets_column = [&] {
+        cudf::column_view offsets_view( cudf::data_type{cudf::INT32}, strings_count+1, 
+                                        strings.offsets().data<int32_t>() );
+        auto table_offsets = cudf::experimental::detail::copy_if( cudf::table_view({offsets_view}),
+                [d_strings, strings_count] __device__ (cudf::size_type idx) {
+                    if( idx==strings_count )
+                        return true;
+                    if( d_strings.is_null(idx) )
+                        return false;
+                    return !d_strings.element<cudf::string_view>(idx).empty();
+                }, mr, stream )->release();
+        strings_count = table_offsets.front()->size()-1;
+        return std::move(table_offsets.front());
+    } (); // this allows freeint the temporary table_offsets
+
+    if( strings_count==0 ) // if no non-empty strings, return an empty column
+        return cudf::make_empty_column(cudf::data_type{cudf::STRING});
+    CUDF_EXPECTS( strings_count >= ngrams , "Insufficient number of strings to generate ngrams");
+    // create a temporary column view from the non-empty offsets and chars column views
+    cudf::column_view strings_view(cudf::data_type{cudf::STRING}, strings_count, 
+                                   nullptr, nullptr, 0, 0,
+                                   {non_empty_offsets_column->view(), strings.chars()});
+    strings_column = cudf::column_device_view::create(strings_view,stream);
+    d_strings = *strings_column;
+
+    // compute the number of strings of ngrams
+    auto ngrams_count = strings_count - ngrams + 1;
+
+    // build output offsets by computing the output bytes for each generated ngram
+    auto offsets_transformer_itr = thrust::make_transform_iterator( thrust::make_counting_iterator<cudf::size_type>(0),
+        ngram_generator_fn{d_strings,ngrams,d_separator} );
+    auto offsets_column = cudf::strings::detail::make_offsets_child_column(offsets_transformer_itr,
+                                       offsets_transformer_itr + ngrams_count,
+                                       mr, stream);
+    auto d_offsets = offsets_column->view().data<int32_t>();
+
+    // build the chars column
+    // generate the ngrams from the input strings and copy them into the chars data buffer
+    cudf::size_type total_bytes = thrust::device_pointer_cast(d_offsets)[ngrams_count];
+    auto chars_column = cudf::strings::detail::create_chars_child_column( ngrams_count, 0, total_bytes, mr, stream );
+    auto d_chars = chars_column->mutable_view().data<char>();
```

Comment:
I think this one too
```suggestion
    auto const d_chars = chars_column->mutable_view().data<char>();
```

Reply:
This is a mutable column data pointer and should not be const.

Reply:
From what I see, the data is mutable, but the pointer itself could be const. The data does not change the address.

Reply:
I see what you mean. I that case, I'd rather remove the auto to make it clearer and use this instead.
```
char* const d_chars = chars_column->mutable_view().data<char>();
```